libsolidity/codegen/ir/IRGenerate.cpp

```diff
@@ -0,0 +1,212 @@
+	m_body = make_shared<assembly::Block>();
+	m_body->location = _contract.location();
+	m_contracts[_contract.fullyQualifiedName()] = m_body;
+
+	ASTNode::listAccept(_contract.definedFunctions(), *this);
+
+	buildDispatcher(_contract);
+
+	return false;
+}
+
+void IRGenerate::buildDispatcher(ContractDefinition const& _contract)
+{
+	appendFunction(R"(
+	{
+		// Revert if value was received.
+		function ensureNoValueTransfer()
+		{
+			switch callvalue()
+			case 0:u256 {}
+			default { revert(0:u256, 0:u256) }
+		}
+
+		// Extract 32 bit method identifier
+		function extractCallSignature() -> sig:u256
+		{
+			// FIXME: replace with constant
+			sig := div(calldataload(0:u256), exp(2:u256, 224:u256))
+		}
+	}
+	)");
+
+	assembly::Switch _switch;
+	_switch.expression = make_shared<assembly::Statement>(createFunctionCall("extractCallSignature"));
+
+	for (auto const& function: _contract.definedFunctions())
+	{
+		if (!function->isPartOfExternalInterface())
+			continue;
+
+		assembly::Literal literal;
+		literal.kind = assembly::LiteralKind::Number;
+		literal.value = toHex(FixedHash<4>::Arith(FixedHash<4>(dev::keccak256(function->externalSignature()))), HexPrefix::Add);
+		literal.type = "u256";
+
+		assembly::Block body;
+		if (!function->isPayable())
+			body.statements.emplace_back(createFunctionCall("ensureNoValueTransfer"));
+		body.statements.emplace_back(createFunctionCall("_" + function->name()));
+
+		assembly::Case _case;
+		_case.value = make_shared<assembly::Literal>(literal);
+		_case.body = std::move(body);
+
+		_switch.cases.emplace_back(_case);
+	}
+
+	assembly::Case defaultCase;
+	if (auto const& fallbackFunction = _contract.fallbackFunction())
+	{
+		assembly::Block body;
+		if (!fallbackFunction->isPayable())
+			body.statements.emplace_back(createFunctionCall("ensureNoValueTransfer"));
+		body.statements.emplace_back(createFunctionCall("fallback"));
+		defaultCase.body = std::move(body);
+	}
+	else
+		defaultCase.body = wrapInBlock(createRevert());
+	_switch.cases.emplace_back(defaultCase);
+
+	m_body->statements.emplace_back(_switch);
+}
+
+bool IRGenerate::visit(FunctionDefinition const& _function)
+{
+	solUnimplementedAssert(_function.isImplemented(), "Unimplemented functions not supported yet.");
+	solUnimplementedAssert(_function.modifiers().empty(), "Modifiers not supported yet.");
+	solUnimplementedAssert(_function.parameters().empty(), "Parameters not supported yet.");
+	solUnimplementedAssert(_function.returnParameters().empty(), "Return parameters not supported yet.");
+
+	assembly::FunctionDefinition funDef;
+	if (_function.name().empty())
+		funDef.name = "fallback";
+	else
+		funDef.name = "_" + _function.name();
+	funDef.location = _function.location();
+	m_currentFunction = funDef;
+	_function.body().accept(*this);
+	return false;
+}
+
+void IRGenerate::endVisit(FunctionDefinition const&)
+{
+	// invalidate m_currentFunction
+	m_body->statements.emplace_back(m_currentFunction);
```

Comment:
I think it would be much more convenient if we had a visitor that would return an AST structure or a string directly.